### PR TITLE
Static linking and a GNU Make template

### PR DESCRIPTION
**Pull request by nelsonje**
@bholt, check this out. I converted almost all our third-party libraries to link statically, removing our dependence on RPATH/RUNPATH. CMake still generates an RPATH, but that's apparently it's default to aid in development. If we add an install target, we could remove our custom configuration and it would strip out the RPATH in the installed binary.

The Boost unit test library is the one exception; we need it to be dynamic because we use its implicit main() feature.

The other piece here is the GNU Make include file generator I was telling you about. It's not perfect; I couldn't figure out how to use generator expressions so I generate part of the data through string processing and the rest I hard-code into the makefile template. I think it'll be good enough for now.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -254,13 +273,21 @@ else()
...
+  set(Boost_LIBRARY_DIR ${THIRD_PARTY_ROOT}/lib)
   list(APPEND tool_list third-party-boost)
 endif()
 
+# we don't use ${Boost_LIBRARIES} because we want to link the
+# following two libraries statically for all targets, and the unit
+# test library dynamically, but only on test targets.
+list(APPEND GRAPPA_LIBS
+  libboost_system.a
+  libboost_filesystem.a)
```
Comment (bholt) [inline CMakeLists.txt]: what if we found the multi-threaded version of Boost? Won't we need `libboost_system-mt.a`?

File: util/grappa.mk [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+# This uses Make's implicit rules, so "make grappa_app" will look for
+# grappa_app.cpp, compile it into grappa_app.o, and link grappa_app.o
+# into a binary named grappa_app.
+
+# The GNU Make implicit compilation recipe for C++ files which turns a
+# file n.cpp/n.cc/n.C into n.o looks like: 
+# $(CXX) $(CPPFLAGS) $(CXXFLAGS) -c -o n.o n.cpp
+
+# The GNU Make implicit link recipe that turns n.o into an executable n is:
+# $(CC) $(LDFLAGS) n.o $(LOADLIBES) $(LDLIBS) -o n
```
Comment (bholt) [inline util/grappa.mk]: is it actually `LOADLIBS`?

**Comment by bholt** [comment]
Cool, I think static linking makes perfect sense in this case.

Any idea if the travis ci build as anything to do with your changes?

File: util/grappa.mk (hunk 2 above)
Comment (nelsonje) [inline util/grappa.mk]: Curiously, no: https://www.gnu.org/software/make/manual/html_node/Catalogue-of-Rules.html#Catalogue-of-Rules

**Pull request merged**